### ISSUE TITLE
should not use replacer at caddyfile parse time

### ISSUE THREAD

**Issue by elee1766** [pberkel/caddy-storage-redis#4]
caddy placeholders like {env.VALUE} should not be evaluated when parsing the caddyfile.

the syntax {env.VALUE} is for runtime environmental variables, and so should be preserved as strings in configuration

the syntax {$VALUE} will result in environmental variables substituted at caddyfile parse time. this is done by the caddyfile parser and you dont need to do anything to get the behavior.

i will try to make a pr fixing this, aka changing it so that the runtime placeholders values are parsed in provision.

this behavior breaks the flow of caddyfile -> json -> caddy, which is critical for validation and doing things like hot reload.

**Comment by pberkel** [comment]
Hi @elee1766 thanks for your comments and contribution, the changes you are proposing seem reasonable, but can you provide a concrete example of the specific problems you encounter with the caddyfile -> json -> caddy flow so I can better understand the issue? Thanks!

**Comment by elee1766** [comment]
hi, yeah here's a very simple example @pberkel 

imagine a config like this:
```
{
	admin off
	storage redis {
		host "{env.REDIS_HOST}"
	}
}

:3838 {
}
```

If i run caddyfile adapt on this to get json, and we use the existing behavior, it will actually fail to generate a json. the reason is because the place i am running caddyfile adapt does NOT have `REDIS_HOST` env var set. As a result, it tries to validate the empty string, and fails

This is a problem, because I can't even pass this json to my runtime environment where i DO have `REDIS_HOST` env var set. importantly, the place i run adapt is a completely different environment from the place i run caddy.

After the patch, i am able to produce a config that looks like this with adapt, which then will function properly at runtime on the other host:

```
{
  "admin": {
    "disabled": true
  },
  "logging": {
    "logs": {
      "default": {
        "level": "DEBUG"
      }
    }
  },
  "storage": {
    "address": null,
    "client_type": "simple",
    "compression": false,
    "db": 0,
    "encryption_key": "",
    "host": [
      "{env.REDIS_HOST}"
    ],
    "key_prefix": "caddy",
    "master_name": "",
    "module": "redis",
    "password": "",
    "port": [
      "6379"
    ],
    "route_by_latency": false,
    "route_randomly": false,
    "timeout": "",
    "tls_enabled": false,
    "tls_insecure": true,
    "tls_server_certs_path": "",
    "tls_server_certs_pem": "",
    "username": ""
  },
  "apps": {
    "http": {
      "servers": {
        "srv0": {
          "listen": [
            ":3838"
          ]
        }
      }
    }
  }
}
```


let me know if this doesn't make sense.

**Comment by pberkel** [comment]
Ok I now better understand the subtle differences between evaluating environment variables during caddyfile parsing verses at runtime. Let's work towards getting your PR merged into the codebase. Thanks!

**Comment by pberkel** [comment]
Thanks for your contribution @elee1766 your fix has been merged and tagged release [v1.2.0](https://github.com/pberkel/caddy-storage-redis/tree/v1.2.0).
